daft/io/sql/sql_data_sink.py

```diff
@@ -0,0 +1,339 @@
+"""SQL Data Sink for writing DataFrames to SQL databases."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Callable, Literal, TypedDict
+
+from sqlalchemy import (
+    JSON,
+    BigInteger,
+    Boolean,
+    Column,
+    Date,
+    DateTime,
+    Float,
+    Integer,
+    LargeBinary,
+    MetaData,
+    String,
+    Table,
+    Time,
+    create_engine,
+    inspect,
+    text,
+)
+
+from daft.datatype import DataType
+from daft.dependencies import pa
+from daft.io import DataSink
+from daft.io.sink import WriteResult
+from daft.recordbatch.micropartition import MicroPartition
+from daft.schema import Schema
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+    from sqlalchemy.engine import Connection
+    from sqlalchemy.types import TypeEngine
+
+
+__all__: tuple[str, ...] = (
+    "SQL_SINK_MODES",
+    "SQLDataSink",
+    "WriteSqlResult",
+)
+
+SQL_SINK_MODES = Literal["append", "overwrite"]
+"""The supported write modes for the SQLDataSink.
+
+The supported modes are:
+    - "append": Inserts new rows into an existing table. It's an error if the table does not exist.
```

Comment:
I believe in the rest of the code base we simply create the table if it does not already exist.